panelstudio-mc19/src/main/java/com/lukflug/panelstudio/mc17/MinecraftGUI.java

```diff
@@ -0,0 +1,223 @@
+		getGUI().exit();
+	}
+	
+	@Override
+	public void render (MatrixStack matrices, int mouseX, int mouseY, float partialTicks) {
+		matrixStack=matrices;
+		mouse=new Point((int)Math.round(mouseX),(int)Math.round(mouseY));
+		renderGUI();
+	}
+	
+	@Override
+	public boolean mouseScrolled (double mouseX, double mouseY, double scroll) {
+		if (!super.mouseScrolled(mouseX,mouseY,scroll)) {
+			mouse=new Point((int)Math.round(mouseX),(int)Math.round(mouseY));
+			if (scroll!=0) {
+				if (scroll>0) getGUI().handleScroll(-getScrollSpeed());
+				else getGUI().handleScroll(getScrollSpeed());
+			}
+		}
+		return true;
+	}
+
+	@Override
+	public boolean mouseClicked (double mouseX, double mouseY, int clickedButton) {
+		if (!super.mouseReleased(mouseX,mouseY,clickedButton)) {
+			mouse=new Point((int)Math.round(mouseX),(int)Math.round(mouseY));
+			switch (clickedButton) {
+				case IInterface.LBUTTON -> lButton = true;
+				case IInterface.RBUTTON -> rButton = true;
+			}
+			getGUI().handleButton(clickedButton);
+		}
+		return true;
+	}
+
+	@Override
+	public boolean mouseReleased (double mouseX, double mouseY, int releaseButton) {
+		if (!super.mouseReleased(mouseX,mouseY,releaseButton)) {
+			mouse=new Point((int)Math.round(mouseX),(int)Math.round(mouseY));
+			switch (releaseButton) {
+				case IInterface.LBUTTON -> lButton = false;
+				case IInterface.RBUTTON -> rButton = false;
+			}
+			getGUI().handleButton(releaseButton);
+		}
+		return true;
+	}
+	
+	@Override
+	public boolean keyPressed (int keyCode, int scanCode, int modifiers) {
+		this.modifiers=modifiers;
+		if (keyCode==GLFW.GLFW_KEY_ESCAPE) exitGUI();
+		else getGUI().handleKey(keyCode);
+		return true;
+	}
+	
+	@Override
+	public boolean charTyped (char chr, int modifiers) {
+		this.modifiers=modifiers;
+		getGUI().handleChar(chr);
+		return true;
+	}
+	
+	@Override
+	public boolean shouldPause() {
+		return false;
+	}
+	
+	/**
+	 * Get the {@link GUI} to be rendered.
+	 * @return current GUI
+	 */
+	protected abstract GUI getGUI();
+	
+	/**
+	 * Get current {@link GUIInterface}.
+	 * @return the current interface
+	 */
+	protected abstract GUIInterface getInterface();
+	
+	/**
+	 * Get current scroll speed.
+	 * @return the scroll speed
+	 */
+	protected abstract int getScrollSpeed();
+	
+	
+	/**
+	 * Implementation of {@link GLInterface} to be used with {@link MinecraftGUI}
+	 * @author lukflug
+	 */
+	public abstract class GUIInterface extends GLInterface {
+		/**
+		 * Constructor.
+		 * @param clipX whether to clip in the horizontal direction
+		 */
+		public GUIInterface (boolean clipX) {
+			super(clipX);
+		}
+		
+		@Override
+		public boolean getModifier (int modifier) {
+			return switch (modifier) {
+				case SHIFT -> (modifiers & GLFW.GLFW_MOD_SHIFT) != 0;
+				case CTRL -> (modifiers & GLFW.GLFW_MOD_CONTROL) != 0;
+				case ALT -> (modifiers & GLFW.GLFW_MOD_ALT) != 0;
+				case SUPER -> (modifiers & GLFW.GLFW_MOD_SUPER) != 0;
+				default -> false;
+			};
```

Comment:
Please keep old C-style switch cases for consistency with the rest of the codebase.